x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/downsample/DataStreamTests.java

```diff
@@ -0,0 +1,248 @@
+        final RolloverResponse rolloverResponse = client().admin().indices().rolloverIndex(new RolloverRequest(dataStreamName, null)).get();
+        // NOTE: here we calculate a delay to index documents because the next data stream write index is created with a start time of
+        // (about) two hours in the future. As a result, we need to have documents whose @timestamp is in the future to avoid documents
+        // being indexed in the old data stream backing index.
+        final String newIndexStartTime = client().admin()
+            .indices()
+            .prepareGetSettings(rolloverResponse.getNewIndex())
+            .get()
+            .getSetting(rolloverResponse.getNewIndex(), IndexSettings.TIME_SERIES_START_TIME.getKey());
+        indexDocs(dataStreamName, 10, Instant.parse(newIndexStartTime).toEpochMilli());
+        client().admin()
+            .indices()
+            .updateSettings(
+                new UpdateSettingsRequest().indices(rolloverResponse.getOldIndex())
+                    .settings(Settings.builder().put(IndexMetadata.SETTING_BLOCKS_WRITE, true).build())
+            )
+            .actionGet();
+
+        // WHEN (simulate downsampling as done by an ILM action)
+        final String downsampleTargetIndex = DataStream.BACKING_INDEX_PREFIX + dataStreamName + "-downsample-1h";
+        final DownsampleAction.Request downsampleRequest = new DownsampleAction.Request(
+            rolloverResponse.getOldIndex(),
+            downsampleTargetIndex,
+            new DownsampleConfig(DateHistogramInterval.HOUR)
+        );
+        final AcknowledgedResponse downsampleResponse = client().admin()
+            .indices()
+            .execute(DownsampleAction.INSTANCE, downsampleRequest)
+            .actionGet();
+
+        /*
+         * Force an index update to avoid failing with "Index updates are expected as index settings version has changed",
+         * due to a possible bug while checking settings versions and actual settings/metadata changes.
+         * See {@link IndexSettings#updateIndexMetadata}.
+         */
+        client().admin()
+            .indices()
+            .updateSettings(
+                new UpdateSettingsRequest().indices(downsampleTargetIndex)
+                    .settings(Settings.builder().put(IndexMetadata.SETTING_INDEX_HIDDEN, false).build())
+            )
+            .actionGet();
+
+        final ModifyDataStreamsAction.Request modifyDataStreamRequest = new ModifyDataStreamsAction.Request(
+            List.of(
+                DataStreamAction.removeBackingIndex(dataStreamName, rolloverResponse.getOldIndex()),
+                DataStreamAction.addBackingIndex(dataStreamName, downsampleTargetIndex)
+            )
+        );
+        client().execute(ModifyDataStreamsAction.INSTANCE, modifyDataStreamRequest).actionGet();
+
+        // THEN
+        assertThat(downsampleResponse.isAcknowledged(), equalTo(true));
```

Comment:
Maybe you should move this assertion immediately after the downsample action is invoked. Here, it feels a bit off place. 

Alternatively, you can drop this assertion as none of the other actions are being asserted.

Reply:
I placed it there since it is after the "THEN" because i one of the things to verify. When I write such tests I always try to separate the "GIVEN", "WHEN" and "THEN" sections in such a way to have assertions all grouped in the "THEN" section, avoiding mixing actions with the verification of the outcome.